pkg/virt-handler/vm.go

```diff
@@ -187,3 +203,224 @@ func MapPersistentVolumes(vm *v1.VM, restClient cache.Getter, namespace string)
+					logging.DefaultLogger().Object(vm).Info().Msgf("Got result from IPAM plugin: %v", result)
+
+					cniConf := NetConf{}
+					if err := json.Unmarshal(conf.Bytes, &cniConf); err != nil {
+						return vm, fmt.Errorf("failed to load netconf: %v", err)
+					}
+
+					calicoClient, err := CreateClient(cniConf)
+					if err != nil {
+						return vm, err
+					}
+
+					// Create the endpoint object and configure it.
+					var endpoint *api.WorkloadEndpoint
+					endpoint = api.NewWorkloadEndpoint()
+					endpoint.Metadata.Name = "eth0"
+					endpoint.Metadata.Node = hostname
+					endpoint.Metadata.Orchestrator = orchestrator
+					endpoint.Metadata.Workload = workload
+					//endpoint.Metadata.Labels = labels
+
+					// Set the profileID according to whether Kubernetes policy is required.
+					// If it's not, then just use the network name (which is the normal behavior)
+					// otherwise use one based on the Kubernetes pod's Namespace.
+					if cniConf.Policy.PolicyType == "k8s" {
+						endpoint.Spec.Profiles = []string{fmt.Sprintf("k8s_ns.%s", namespace)}
+					} else {
+						endpoint.Spec.Profiles = []string{cniConf.Name}
+					}
+
+					// Populate the endpoint with the output from the IPAM plugin.
+					if err = PopulateEndpointNets(endpoint, result); err != nil {
+						// Cleanup IP allocation and return the error.
+						ReleaseIPAllocation(logger, cniConf.IPAM.Type, conf.Bytes)
+						return vm, err
+					}
+					logging.DefaultLogger().Object(vm).Info().Msgf("Populated endpoint: %v", endpoint)
+
+					// create a tap device
+					nameTemplate := fmt.Sprintf("tap-%s-%%d", vmCopy.Spec.Domain.UUID[0:8])
+					ifName, err := tuntap.CreatePersistentIface(nameTemplate, tuntap.Tap)
+					if err != nil {
+						return vm, fmt.Errorf("Failed to create a tap device: %v", err)
+					}
+
+					link, err := netlink.LinkByName(ifName)
+					if err != nil {
+						return vm, fmt.Errorf("cannot find link %q", ifName)
+					}
+					logging.DefaultLogger().Object(vm).Info().Msgf("Link(tap) %v created", link)
+
+					if err = netlink.LinkSetUp(link); err != nil {
+						return nil, fmt.Errorf("cannot set link up %q", ifName)
+					}
+
+					mac := link.Attrs().HardwareAddr
+					endpoint.Spec.MAC = &cnet.MAC{HardwareAddr: mac}
+					endpoint.Spec.InterfaceName = ifName
+					logger.WithField("endpoint", endpoint).Info("Added Mac and interface name to endpoint")
+
+					// Write the endpoint object (either the newly created one, or the updated one)
+					if _, err := calicoClient.WorkloadEndpoints().Apply(endpoint); err != nil {
+						// Cleanup IP allocation and return the error.
+						ReleaseIPAllocation(logger, cniConf.IPAM.Type, conf.Bytes)
+						return vm, err
+					}
+					logger.Info("Wrote updated endpoint to datastore")
+
+					logging.DefaultLogger().Object(vm).Info().Msgf("Mapping %s", ifName)
+					newIface := v1.Interface{}
+
+					newIface.Type = "ethernet"
+					newIface.MAC = new(v1.MAC)
+					newIface.MAC.MAC = mac.String()
```

Comment:
1. MapInterfaces() function
  a) gets an IP address from calico-ipam.
  b) creates a tap device and get a MAC address of it.
  c) creates a calico endpoint with [IP,  MAC] information in the calico data store. (calico uses etcd as the data store.)

2. Calico add a route to the tap device with the IP address.

3. `[Calico-DHCP](https://github.com/openstack/networking-calico/tree/master/networking_calico/agent)` will watch the calico data store and upate `dnsmasq` with the updated endpoint data.